### ISSUE TITLE
"The model for interpretation of GL commands is client-server..."

### ISSUE THREAD

**Issue by arielm** [kripken/worker-ui#5]
Quoted from the [OpenGL 1.1 specs](https://www.opengl.org/documentation/specs/version1.1/glspec1.1/node10.html):

_The model for interpretation of GL commands is client-server. That is, a program (the client) issues commands, and these commands are interpreted and processed by the GL (the server). The server may or may not operate on the same computer as the client. In this sense, the GL is 'network-transparent.'' A server may maintain a number of GL contexts, each of which is an encapsulation of current GL state. A client may choose to connect to any one of these contexts. Issuing GL commands when the program is not connected to a context results in undefined behavior._

A few years later, it seems that this approach is still relevant, e.g, in projects like [Wayland](http://wayland.freedesktop.org/architecture.html).

In short: Wayland is based on EGL / GLES2 and it handles input as well (it is one of the supported platforms in recent GLFW versions...)

Open question: how to put all all this wisdom to use with Emscripten and WebGL?

**Comment by arielm** [comment]
A recap of the _state-of-the-art_, as I understand it:

1) The _async command-queue_ approach:
- Today: [WebGL in Web Workers, Today – and Faster than Expected!](https://blog.mozilla.org/research/2014/07/22/webgl-in-web-workers-today-and-faster-than-expected)
- Future: [WebVulkan and multithreaded command queue assembly](https://www.khronos.org/webgl/public-mailing-list/archives/1503/msg00001.html)
- Should help:
  - SharedArrayBuffer support in Firefox nightlies
  - New experimental support for [pthreads in Emscripten](https://groups.google.com/forum/#!topic/emscripten-discuss/gQQRjajQ6iY)

2) The _shared frame-buffer_ approach:
- Big-picture: the [Wayland](http://wayland.freedesktop.org/architecture.html) architecture
- Future:
  - The [OffscreenCanvas](https://wiki.whatwg.org/wiki/OffscreenCanvas) proposal
  - Some context: [Canvas in workers](http://lists.w3.org/Archives/Public/public-whatwg-archive/2013Oct/0152.html)
- Should help:
  - Some recent [Mozilla patches](https://bugzilla.mozilla.org/show_bug.cgi?id=709490)?
